src/board/system76/common/power.c

```diff
@@ -334,13 +334,42 @@ void power_off(void) {
     update_power_state();
 }
 
-// Set the CPU power limit appropriately
-static bool power_peci_limit(bool ac) {
+bool power_peci_limit(bool ac) {
     if (peci_available()) {
-        uint16_t watts = ac ? POWER_LIMIT_AC : POWER_LIMIT_DC;
-        // Set PL4 using PECI
-        int16_t res = peci_wr_pkg_config(60, 0, ((uint32_t)watts) * 8);
-        DEBUG("power_peci_limit %d = %d\n", watts, res);
+        uint32_t watts;
+
+        if (ac) {
+            // AC adapter rating with a 15% margin for conversion  losses & spikes
+            // This results in average pwr consumption ~equal to adapter rating
+            // TODO handle this differently for models with HPB enabled
+            watts = (uint32_t)battery_charger_input_current *
+                (uint32_t)battery_charger_input_voltage * 85 / 100000;
+        } else {
+            watts = POWER_LIMIT_DC;
+        }
+
+        uint8_t retry = 10;
+        uint16_t res;
+        do {
+            res = peci_wr_pkg_config(PECI_REG_PKG_CFG_PSYS_PL2, 0, PECI_PSYS_PL2(watts));
+            delay_us(1);
+        } while (res != 0x40 && retry--);
+        DEBUG("PECI SET PsysPL2 %ld = %d\n", watts, res);
+
+        retry = 10;
+        do {
+            res = peci_wr_pkg_config(PECI_REG_PKG_CFG_PL3, 0, PECI_PL3(watts, 6, 4));
```

Comment:
PL3 should be 2W higher than PL2 per Intel recommendations

Reply:
PL3 and PsysPL3 share the same register, if Psys is enabled then this register is used for PsysPL3 and not PL3 (doc #613972)

Reply:
Yes, it is recommended as:

PL3 = PL2 + 2W
PsysPL3 = PsysPL2 + 2W

Reply:
done